### PR TITLE
Remove PooledClass from isomorphic build

### PR DESCRIPTION
**Pull request by gaearon**
This inlines `traverseAllChildren` logic into `ReactChildren` since it’s now the only consumer. Then I change `forEach` and `map` to use the same bookkeeping fields. This lets me unroll the `PooledClass` abstraction into a single specific implementation, and thus remove it from the isomorphic build.

See individual commits.

### CODE REVIEW

File: src/isomorphic/children/ReactChildren.js [hunk 1]
```diff
@@ -164,7 +353,7 @@ function mapChildren(children, func, context) {
   return result;
 }
 
-function forEachSingleChildDummy(traverseContext, child, name) {
```
Comment (aweary) [inline src/isomorphic/children/ReactChildren.js]: We're already importing the `emptyFunction` library in `ReactChildren`, could use just use `emptyFunction.thatReturnsNull` and get rid of this?

File: scripts/rollup/results.json [hunk 2]
```diff
@@ -1,28 +1,28 @@
...
-      "size": 56905,
-      "gzip": 14622
+      "size": 52704,
+      "gzip": 13552
     },
     "React-prod.js (FB_PROD)": {
-      "size": 27099,
-      "gzip": 7215
+      "size": 24229,
+      "gzip": 6612
```
Comment (iamdustan) [inline scripts/rollup/results.json]: whoa. why is FB_PROD so much larger than NODE_PROD/UMD_PROD? Legacy files that are being consumed internally yet?

Comment (gaearon) [inline scripts/rollup/results.json]: No, it's just not minified. We might minify it eventually but for now it doesn't seem necessary (it goes through our pipeline anyway), and it's easier to see what happens during the sync.

Comment (iamdustan) [inline scripts/rollup/results.json]: Ah! Great point. :thumbsup:

File: src/isomorphic/children/ReactChildren.js (hunk 1 above)
Comment (gaearon) [inline src/isomorphic/children/ReactChildren.js]: Addressed.

File: src/isomorphic/children/ReactChildren.js [hunk 3]
```diff
@@ -11,39 +11,251 @@
...
+var SEPARATOR = '.';
+var SUBSEPARATOR = ':';
+
+/**
+ * Escape and wrap key so it is safe to use as a reactid
+ *
+ * @param {string} key to be escaped.
+ * @return {string} the escaped key.
+ */
+function escape(key: string): string {
```
Comment (VladBlow) [inline src/isomorphic/children/ReactChildren.js]: Hey,
Why you added types in flow style if this file have't flow comment?

Comment (gaearon) [reply src/isomorphic/children/ReactChildren.js]: Probably accidentally or by copy paste. Want to send a PR to Flow-ify it?

Comment (VladBlow) [reply src/isomorphic/children/ReactChildren.js]: yep! I can delete useless flow typechecking or add comment flow. If I add flow comment maybe I can try to update types for all file? If you want this.

Comment (gaearon) [reply src/isomorphic/children/ReactChildren.js]: Sure.

**Pull request merged**